bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/datalayer/ComponentDataImpl.java

```diff
@@ -15,101 +15,331 @@
+                .getLinkUrl()
+                .map(Supplier::get)
+                .orElse(null);
+        }
+        return this.linkUrl;
+    }
+
+    @Override
+    @Nullable
+    public String[] getShownItems() {
+        if (this.shownItems == null) {
+            this.shownItems = this.getDataLayerSupplier()
+                .getShownItems()
+                .map(Supplier::get)
+                .orElse(null);
+        }
+        if (this.shownItems != null) {
+            return Arrays.copyOf(this.shownItems, this.shownItems.length);
+        }
+        return null;
+    }
 
-            if (lastModified == null) {
-                lastModified = valueMap.get(JcrConstants.JCR_CREATED, Calendar.class);
-            }
+    @Override
+    @Nullable
+    public AssetData getAssetData() {
+        if (this.assetData == null) {
+            this.assetData = this.getDataLayerSupplier()
+                .getAssetData()
+                .map(Supplier::get)
+                .orElse(null);
         }
+        return this.assetData;
+    }
 
-        if (lastModified != null) {
-            return lastModified.getTime();
+    @Override
+    @Nullable
+    public String getTemplatePath() {
+        if (this.templatePath == null) {
+            this.templatePath = this.getDataLayerSupplier()
+                .getTemplatePath()
+                .map(Supplier::get)
+                .orElse(null);
         }
+        return this.templatePath;
+    }
 
-        return null;
+    @Override
+    @Nullable
+    public String getLanguage() {
+        if (this.language == null) {
+            this.language = this.getDataLayerSupplier()
+                .getLanguage()
+                .map(Supplier::get)
+                .orElse(null);
+        }
+        return this.language;
     }
 
     @Override
-    public String getText() {
-        return component.getDataLayerText();
+    @Nullable
+    public String[] getTags() {
+        if (this.tags == null) {
+            this.tags = this.getDataLayerSupplier()
+                .getTags()
+                .map(Supplier::get)
+                .orElse(null);
+        }
+        if (this.tags != null) {
+            return Arrays.copyOf(this.tags, this.tags.length);
+        }
+        return null;
     }
 
     @Override
-    public String getLinkUrl() {
-        return component.getDataLayerLinkUrl();
+    @Nullable
+    public String getUrl() {
+        if (this.url == null) {
+            this.url = this.getDataLayerSupplier()
+                .getUrl()
+                .map(Supplier::get)
+                .orElse(null);
+        }
+        return this.url;
     }
 
     @Override
+    @Nullable
     public String getJson() {
+        return ComponentDataImpl.getJson(this);
+    }
+
+    /**
+     * Get the JSOn for component data.
+     *
+     * @param data The component data.
+     * @return The JSON.
+     */
+    @Nullable
+    static String getJson(@NotNull final ComponentData data) {
```

Comment:
Does it make sense to move this logic to a `static`? Were you planning on moving it to a public utils class?

Reply:
This is probably an artifact of refactoring and is no longer needed. 
We could move it to a utility class if there is any purpose to that, otherwise the change can probably just be reverted.

Reply:
we may want to move this method to the new `ComponentUtils` class.